### PR TITLE
Circuit-breaker for suspicions learnt from mergestate

### PR DESCRIPTION
**Pull request by bpineau**
Rate-limiter / circuit-breaker for suspicions received when consul
merges a remote state. Configurable in non-enforcement mode (only
log when limit are exceeded). All 3 new config params are reloadable.

Not doing that change lightly hearted, as that's not something we can
submit upstream. If this is enough to get rid of our issues, we then
would have nailed down a consul fragility that we'll discuss with upstream.

As nobody want to live with a consul fork forever, here are the
details and rationale, for the record:

Probing, overview (and observables):
-----------------------------------

* Every `probe_interval` (default 500ms) each node "pings" (over udp, then tcp fallback) an other node chosen at random
* If the other node fails to pass the probe, the probing node will enqueue a "suspectMsg" for gossip broadcast and log "Suspect %s has failed, no acks received"
* The broadcast queue's gossip messages are sent (up to 1400 bytes of them) every 200ms to 3 random other nodes, and remains in queue until sent 20 times (x3)
* Every node will relay gossip messages they see for the first time (enqueue for broadcast, send 20 (x3) times, etc)
* If a live node realize being suspected (from a received "suspectMsg"), it will enqueue an "aliveMsg" (refutation) for broadcast
* When a node enqueue a refutation (aliveMsg) for broadcast, it also logs the node it received the "suspectMsg" from ("Refuting a suspect message (from: %s)")
* After a deadline, and if not refutation was received, the suspecting node will mark the suspected node as dead (enqueue a "deadMsg" for broadcast)
* A "dead" node (whose name is broadcasted in a "deadMsg") is removed from the services it was member of

There's more details (adaptive "anti-entropy" timers, indirect pings, udp then tcp fallback, broadcast queue reordering/prioritisations, ...), but that's the general flow.

Complementary to that mechanism, nodes regularly exchange their state with a peer at random (push/pull), over TCP.
For each death or suspicion they learned from the peer they received a state from, nodes send a suspectMsg.

Consul "blip" observations:
--------------------------

* When a "blip" occurs, we're first seeing a suspectMsg surge (over a thousand in a few seconds), quickly followed by aliveMsg (refutations) then deadMsg spikes
* The burst is sudden, and spreads too fast for our `gossip.memberlist.msg.suspect` metric granularity to unveil a patient zero (if any)
* Those suspectMsg are likely false positives:
  * They are soon followed by refutations (aliveMsg) burst
  * Suspected machines are live nodes that looks otherwise healthy (and whose consul logged activity by the time they were supposed to fail answering health probes)
  * We don't see the corresponding initial health probe failure logs ("Suspect %s has failed"): the origin may not be failing health probes
  * Having hundreds of unrelated nodes/roles frequently failing consul probes -and only consul probes- at the exact same time seems unlikely
  * They seem to originate from unhealthy (scheduled for maintenance) nodes (see below, refutations source)
* Consul memberlist broadcast queue has a limited rate (a node can broadcast up to 1400 byte worth of distinct messages every 200ms)
* aliveMsg are about x3.5 larger than suspectMsg: they'll propagate slower
* During incidents, we're seeing (instrumented) consul queues accumulating thousands of messages: gossip propagation is effectively delayed
* The broadcast queues congestions can explain why refutations don't propagate soon enough to prevent deadMsg and nodes removal from services
* We don't have direct evidence of the suspectMsg source(s), due to abnormally missing "Suspect %s has failed, no acks received" logs, and too coarse metrics granularity
* We see some indirect traces: for every incident, a single node is massively refuted by the others (as seen from "Refuting a suspect message (from: %s)" logs)
* The massively refuted node is not the same machine every time consul blips (but a given machine can be singled out for several blips)
* Most of the time, the massively refuted nodes are also machines "scheduled for maintenance" (due to "degraded hardware") by AWS, sometime exposing network flakiness symptoms
* Aggressively replacing "scheduled for maintenance" nodes seems to calm the blips frequency (but we're far from statistical evidence)
* A single machine can't initiate more than two new suspicions per second: they only probe one node every 500ms (yet we see a single machine refuted thousands time a second)
* A single machine can relay thousands of suspectMsg per second, if received from gossip or pushpull state exchange

In short: we're likely facing a bug (abnormally missing "has failed" logs, etc), triggered by the "degraded hardware" some nodes runs on.
This leads to a spurious suspectMsg burst, causing contentions and "consul blips".

We'd like to better understand how a "degraded hardware" can briefly corrupt a consul to the point it sends suspicions burst.
In any cases, we shouldn't let possibly unhealthy instances broadcast hundreds (or thousands) suspectMsg in a few seconds,
as the cluster can't cope with that volume, and will degenerate in failing many members, escalating to services failures.

### CODE REVIEW

File: vendor/github.com/hashicorp/memberlist/config.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ import (
 	"log"
```
Comment (LeoCavaille) [inline vendor/github.com/hashicorp/memberlist/config.go:5]: I had forked the deps and committed a change there rather than referencing the old revisions in  go.mod, not sure what's best but we should likely be consistent in the fork.

**Comment by bpineau** [comment]
A precision, as that may not be clear in the diff:

For now that patch only rate limit suspectMsg generated during mergeState (when an agent sync with another).
It does not limit new suspicion generated locally (should be low volume anyway, timer based), and does not limit suspectMsg received over broadcast.

We're seeing a node being massively refuted (is designed by the `From` field in false positive suspectMsg).

A node will keep the original From field for suspectMsg received over broadcast, but will update it with his own address for suspectMsg generated from a mergeState.

So the theory for rate limiting mergeState only is that those suspicions really come from an unhealthy node after it merged a corrupted state, as it's unlikely that massively refuted node could have local
y generated that huge amount of suspectMsg in a short timespan. Then installing a patched consul on those nodes (only) would be enough to prevent blips.

Limiting the later (relaying suspectMsg received over broadcast, in addition to those from mergeState) would be easy, and would effectively circuit break a blip while it start propagating, but it's a bit more dangerous as that may also hide real network partitions.

What do you think?

**Review by LeoCavaille — APPROVED** [review]

**Comment by LeoCavaille** [comment]
This initial approach sounds valid to me and I agree that limiting at a later time is a bit dangerous. We can reevaluate depending on how we see the cluster react to this config ^

File: vendor/github.com/hashicorp/memberlist/config.go (hunk 1 above)
Comment (bpineau) [reply vendor/github.com/hashicorp/memberlist/config.go:5]: Ok, did the same for `serf`and `memberlist` (for the later, also fetched changes from 1e67f4113c8b27665ee18f3e74575569364ac7ae , as this commit's change to vendors.json didn't match what was really vendored).
